### DISCUSSION TITLE
How to construct a sketch using the edga-based API and then extrude it in a empty workplane.

### DISCUSSION THREAD

**Discussion by ghost** [CadQuery/cadquery#1374] [Q&A] [answered]
I checked CadQuery's read the docs (https://cadquery.readthedocs.io) for working with sketches.  Perhaps I missed something, but I would like to extrude a sketch to get my **first** 3D object.  From the examples I see how to add a sketch to an existing face, however I am running into road blocks when trying to extrude a sketch off the default workplane.

Here is my example sketch:
```
    r1 = 3/16
    r2 = 2*r1
    h = 2
    b = 1
    angle = 30 # degrees

    result = (
        cq.Sketch()
        .segment((0,0),(0,h))
        .segment(r1,0)
        .segment(h,-90)
        .arc((r2,0), r1, 180, angle)
        .segment(b, 270+angle)
        .segment(r1,180+angle)
        .segment(b,90+angle)
        .arc((r2,0), r2, 180+angle, -angle)
        .close()
    )
```
I have tried a several things so far including:
```
   result = (
        cq.Workplane("XY")
        .sketch()
...
        .extrude(10)
    )
```
This leads to the error 'Sketch; object has no attribute 'extrude'.  Since the workplane does not yet have a face, I cannot use `.faces('>Z')` to build my sketch.  I also tried `.placeSketch(result)`, but I hit a similar error.  I see that one can transform a workplane.  Nevertheless sketching seems to require a face or something _concrete_ to build off in order to extrude.

I would be grateful for guidance.

**Comment by jmwright** [comment]
Here's an example I have in my notes that seems to work.

```
import cadquery as cq

s1 = (
    cq.Sketch()
    .rect(30, 40)
    .rect(10, 10, mode="s", tag="rectcut")
    .vertices(">X", tag="rectcut")
    .fillet(1)
    .vertices("<X", tag="rectcut")
    .fillet(2)
)

r = cq.Workplane().placeSketch(s1).extrude(10)
```

**Reply by ghost** [reply]
Thank you for the quick reply.  I tried something similar in my own attempts, which after testing your solution to my sketch, leads to `ValueError: No pending wires present`.  To summarize @lorenzncode's explanation as to why it resulted in a `ValueError`: The call to `.rect(...)` constructs a face.  My sketch does not create a face.

**Comment by lorenzncode** [comment, marked as answer]
Remove `close()` (because the final arc already closes the edges).
Add `assemble()`.


```py 
import cadquery as cq

r1 = 3/16
r2 = 2*r1
h = 2
b = 1
angle = 30 # degrees

s = (
    cq.Sketch()
    .segment((0,0),(0,h))
    .segment(r1,0)
    .segment(h,-90)
    .arc((r2,0), r1, 180, angle)
    .segment(b, 270+angle)
    .segment(r1,180+angle)
    .segment(b,90+angle)
    .arc((r2,0), r2, 180+angle, -angle)
    .assemble()
)

result = cq.Workplane().placeSketch(s).extrude(10)
```


See the docs section on the Edge-based API and the notes on `assemble`.
https://cadquery.readthedocs.io/en/latest/sketch.html#edge-based-api

`assemble` is not required in @jmwright's example because it calls `rect` which automatically results in constructing a face (face-based API).

To define the sketch inline (instead of `placeSketch`):

```py
...

result  = (
    cq.Workplane()
    .sketch()
    .segment((0,0),(0,h))
    ...
    .arc((r2,0), r2, 180+angle, -angle)
    .assemble()
    .finalize() # finish the sketch
    .extrude(10)
)
```

**Reply by ghost** [reply]
I tested both solutions (Using `placeSketch(s)` & inline).  Both were successful.  Thank you!

Also, I appreciate your advice on `assemble`.  To quote [https://cadquery.readthedocs.io/en/latest/sketch.html#edge-based-api]()  "Once the construction is finished it has to be converted to the face-based representation using `assemble()`. Afterwards, face based operations can be applied."

Thank you also for the fast reply.

**Comment by adam-urbanczyk** [comment]
@Willy-Schnackenpfefferhausen also read this section of the docs https://cadquery.readthedocs.io/en/latest/sketch.html#workplane-integration .

**Comment by ghost** [comment]
Thank you for the suggestion.

In each of the cases that are presented in #workplane-integration, the code begins with an existing face.  I began by using these examples but unfortunately overlooked the answer to my question, which was addressed in #edge-based-api on the same page.  The necessary step for my situation was to use `assemble()`.

**Reply by adam-urbanczyk** [reply]
AFAICT your question was about beginning with a sketch, not rly about assemble. If this is not correct maybe amend the title.

**Reply by ghost** [reply]
I see from the solution how the title comes into question, because I would have had the same issues trying to extrude this sketch on an existing face.  I did not know that at the time of asking.

The solution did answer my initial question as well, which is how does one begin a model with a sketch.  That is without:
```
cq.Workplane()
.box(5,5,1)
.faces('>Z')
```
From the examples in #workplane-integration I could not find how I could build off an empty workplane.  Each example used `.faces(...)`.

Based on this, what would you propose I change my title to?

**Reply by adam-urbanczyk** [reply]
OK, I'll amend the title myself.

**Reply by adam-urbanczyk** [reply]
I'm not sure that we are on the same page though - to me last example in the sketch docs shows how to place a sketch in an empty workplane.

**Reply by ghost** [reply]
I agree with your title.

"... last example in the sketch docs shows how to place a sketch in an empty workplane."  Fair enough, but I did not know what to do with a sketch in an empty workplane other than to loft.
